### PR TITLE
Added UserClass option to DHCPv4

### PR DESCRIPTION
**Pull request by mooneyow**
Added support for UserClass options for DHCPv4 as per the [RFC](https://tools.ietf.org/html/rfc3004)

### CODE REVIEW

File: dhcpv4/option_userclass.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+			totalLength, len(data))
+	}
+
+	for i := 0; i < totalLength; {
+		ucLen := int(data[i])
+		opaqueDataIndex := i + ucLen + 1
+		if len(data) < opaqueDataIndex {
+			return nil, fmt.Errorf("ParseOptUserClass: short data: less than %d bytes", opaqueDataIndex)
+		}
+		opt.UserClasses = append(opt.UserClasses, data[i+1:opaqueDataIndex])
```
Comment (insomniacslk) [inline dhcpv4/option_userclass.go]: nit: for clarity I prefer that we express the slice indexes in terms of `[something:something+something_else]`, i.e. don't use `opaqueDataIndex`, but just use `i` instead. E.g.
```
base := i + 1
if len(data) < base + ucLen {
    // error
}
opt.UserClasses = append(opt.UserClasses, data[base:base+ucLen])
```

File: dhcpv4/option_userclass.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	if len(data) < totalLength {
+		return nil, fmt.Errorf("ParseOptUserClass: short data: length is %d but got %d bytes",
+			totalLength, len(data))
+	}
+
+	for i := 0; i < totalLength; {
+		ucLen := int(data[i])
+		opaqueDataIndex := i + ucLen + 1
+		if len(data) < opaqueDataIndex {
+			return nil, fmt.Errorf("ParseOptUserClass: short data: less than %d bytes", opaqueDataIndex)
```
Comment (insomniacslk) [inline dhcpv4/option_userclass.go]: also print the actual size, it will simplify troubleshooting:
```
return nil, fmt.Errorf("ParseOptUserClass: short data: %d bytes; want: %d", len(data), opaqueDataIndex)
```

(the "got.. want" comes from Go's code review advice, https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures )

File: dhcpv4/option_userclass_test.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+		4, 't', 'e', 's', 't',
+	}
+	opt, err := ParseOptUserClass(expected)
+	require.NoError(t, err)
+	require.Equal(t, len(opt.UserClasses), 2)
+	require.Equal(t, []byte("linuxboot"), opt.UserClasses[0])
+	require.Equal(t, []byte("test"), opt.UserClasses[1])
+}
+
+func TestParseOptUserClassNone(t *testing.T) {
```
Comment (insomniacslk) [inline dhcpv4/option_userclass_test.go]: also write a function to test zero-length user classes, e.g. when
```
data := []byte{0}
```
and
```
data := []byte{10, 'l', 'i', 'n', 'u', 'x', 'b', 'o', 'o', 't', 0}
```

File: dhcpv4/option_userclass_test.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	_, err := ParseOptUserClass(expected)
+	require.Error(t, err)
+}
+
+func TestParseOptUserClass(t *testing.T) {
+	expected := []byte{
+		77, 10, 9, 'l', 'i', 'n', 'u', 'x', 'b', 'o', 'o', 't',
+	}
+	opt, err := ParseOptUserClass(expected)
+	require.NoError(t, err)
```
Comment (insomniacslk) [inline dhcpv4/option_userclass_test.go]: also require that `opt.Code() == OptionUserClassInformation` and `opt.Length() == 10`

**Comment by mooneyow** [comment]
Updated with requested changes 😁
